### ISSUE TITLE
Integration test topologies assert exactly once, but run with at most once.

### ISSUE THREAD

**Issue by billonahill** [twitter/heron#1561]
We're seeing occasional failures like this in CI. This issue is that `IntegrationTestSpout` now decrements the `tuplesToAck` counter only when tuples are acked ([commit link](https://github.com/twitter/heron/commit/1123f731cb4a399ef4717e1c07dcbd22ccbab37d#diff-92e59551fb6a8e1002ba5c3f20f8f61aL127)). Previously it decremented that counter even when a `fail` call was received. What this implies is that our test topologies are dropping tuples and the spout is re-transmitting them. Why the test topologies are dropping tuples is concerning.

A few options we could pursue:

1. Why are tuples failing, causing them to be re-emitted? That shouldn't be happening in our tests
2. We could change `IntegrationTestSpout` back to decrement upon failure.
3. We could change test assertion to assert `atLeastOnce` instead of `exactlyOnce`.
```
ERROR: Actual result did not match expected result
INFO: Actual result ---------- 
['A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B']
INFO: Expected result ---------- 
['A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'A', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B']
ERROR: Checking result failed for 20161113214723_IntegrationTest_MultiSpoutsMultiTasks_861f6ed7-ca34-47c9-9bde-3ba3a2d802e4 topology :: ('Actual result did not match expected result', None)
```

**Comment by objmagic** [comment]
constantly fails both Travis CI and internal CI now. We should prioritize this.

**Comment by billonahill** [comment]
This ticket is blocked on us reaching an agreement for how we should proceed. I suggest we follow option 1 above, because it's unclear to me why tuples should fail in integration tests.

I would like to hear opinions from others though.

**Comment by billonahill** [comment]
@maosongfu, @kramasamy do you have thoughts re why tuples are being consistently dropped by the integration tests (is this expected?) and if that's something we should fix, or obfuscate by changing to at least once verification?

**Comment by billonahill** [comment]
Relevant classes:
https://github.com/twitter/heron/blob/master/integration-test/src/java/com/twitter/heron/integration_test/core/IntegrationTestSpout.java#L135
https://github.com/twitter/heron/blob/master/integration-test/src/java/com/twitter/heron/integration_test/core/IntegrationTestBolt.java

**Comment by maosongfu** [comment]
Process death could lead to tuples dropping and then re-emitting. In earlier integration tests, we consider process death a kind of failure. But seems this assumption is not longer true.

**Comment by billonahill** [comment]
If a process dies currently and tuples get re-emitted, that would trigger a failure during assertion, since we assert exactly once, not at least once.

These tests fail often though, so if we have frequent process deaths in our integration tests, something is really wrong. Hence my hesitation to change to at least once assertion without first identifying the cause of re-transmits.

**Comment by objmagic** [comment]
resolved by #1830 #1815
